common/pinot/pinotQueryValidator.go

```diff
@@ -0,0 +1,271 @@
+		}
+		buf := sqlparser.NewTrackedBuffer(nil)
+		res := ""
+		// validate where expr
+		if sel.Where != nil {
+			res, err = qv.validateWhereExpr(sel.Where.Expr)
+			if err != nil {
+				return "", &types.BadRequestError{Message: err.Error()}
+			}
+		}
+
+		sel.OrderBy.Format(buf)
+		res += buf.String()
+		return res, nil
+	}
+	return whereClause, nil
+}
+
+func (qv *VisibilityQueryValidator) validateWhereExpr(expr sqlparser.Expr) (string, error) {
+	if expr == nil {
+		return "", nil
+	}
+	buf := sqlparser.NewTrackedBuffer(nil)
+
+	switch expr := expr.(type) {
+	case *sqlparser.AndExpr, *sqlparser.OrExpr:
+		return qv.validateAndOrExpr(expr)
+	case *sqlparser.ComparisonExpr:
+		return qv.validateComparisonExpr(expr)
+	case *sqlparser.RangeCond:
+		expr.Format(buf)
+		return buf.String(), nil
+		//return qv.validateRangeExpr(expr)
+	case *sqlparser.ParenExpr:
+		return qv.validateWhereExpr(expr.Expr)
+	default:
+		return "", errors.New("invalid where clause")
+	}
+}
+
+func (qv *VisibilityQueryValidator) validateAndOrExpr(expr sqlparser.Expr) (string, error) {
+	var leftExpr sqlparser.Expr
+	var rightExpr sqlparser.Expr
+	isAnd := false
+
+	switch expr := expr.(type) {
+	case *sqlparser.AndExpr:
+		leftExpr = expr.Left
+		rightExpr = expr.Right
+		isAnd = true
+	case *sqlparser.OrExpr:
+		leftExpr = expr.Left
+		rightExpr = expr.Right
+	}
+
+	leftRes, err := qv.validateWhereExpr(leftExpr)
+	if err != nil {
+		return "", err
+	}
+
+	rightRes, err := qv.validateWhereExpr(rightExpr)
+	if err != nil {
+		return "", err
+	}
+
+	if isAnd {
+		return fmt.Sprintf("%s and %s", leftRes, rightRes), nil
+	}
+
+	return fmt.Sprintf("(%s or %s)", leftRes, rightRes), nil
+}
+
+func (qv *VisibilityQueryValidator) validateComparisonExpr(expr sqlparser.Expr) (string, error) {
+	comparisonExpr := expr.(*sqlparser.ComparisonExpr)
+
+	colName, ok := comparisonExpr.Left.(*sqlparser.ColName)
+	if !ok {
+		return "", errors.New("invalid comparison expression, left")
+	}
+
+	colNameStr := colName.Name.String()
+
+	if !qv.isValidSearchAttributes(colNameStr) {
+		return "", fmt.Errorf("invalid search attribute %q", colNameStr)
+	}
+
+	// Case1: it is system key
+	// this means that we don't need to change the structure of the query,
+	// just need to check if a value == "missing"
+	if definition.IsSystemIndexedKey(colNameStr) {
+		return qv.processSystemKey(expr)
+	}
+	// Case2: when a value is not system key
+	// This means, the value is from Attr so that we need to change the query to be a Json index format
+	return qv.processCustomKey(expr)
+}
+
+// isValidSearchAttributes return true if key is registered
+func (qv *VisibilityQueryValidator) isValidSearchAttributes(key string) bool {
+	validAttr := qv.validSearchAttributes
+	_, isValidKey := validAttr[key]
+	return isValidKey
+}
+
+func (qv *VisibilityQueryValidator) processSystemKey(expr sqlparser.Expr) (string, error) {
+	comparisonExpr := expr.(*sqlparser.ComparisonExpr)
```

Comment:
nit: for all methods that expect a single type, it would be easier to use it directly instead of asserting to the interface.